module/userExtension/helper.js

```diff
@@ -211,304 +213,401 @@ module.exports = class UserExtensionHelper {
+                  } else if (userRole.programOperation === messageConstants.common.REMOVE_OPERATION) {
+                    if (currentRoleInfoIndex !== -1) {
+                      const currentRoles = existingUserProgramRoleMapping[currentRoleInfoIndex].roles;
+
+                      // Remove specified roles
+                      const rolesToKeep = currentRoles.filter((role) => !newRoles.includes(role));
+
+                      // Emit delete events for removed roles
+                      const rolesToRemove = currentRoles.filter((role) => newRoles.includes(role));
+                      for (const role of rolesToRemove) {
+                        kafkaEventPayloads.push({
+                          userId: userProfile.id,
+                          username: userProfile.username,
+                          programId: programObjectId,
+                          role,
+                          eventType: messageConstants.common.DELETE_EVENT_TYPE,
+                        });
+                      }
+
+                      if (rolesToKeep.length === 0) {
+                        // Remove entire program entry if no roles left
+                        existingUserProgramRoleMapping.splice(currentRoleInfoIndex, 1);
+                      } else {
+                        // Update with remaining roles
+                        existingUserProgramRoleMapping[currentRoleInfoIndex].roles = rolesToKeep;
+                      }
+                    }
+                    // If program doesn't exist, nothing to remove - no error needed
+                  }
+                }
               }
 
-              user = await database.models.userExtension.findOneAndUpdate(
-                {
-                  _id: existingUser._id,
-                },
-                updateQuery,
-                {
-                  new: true,
-                  returnNewDocument: true,
-                },
-              );
+              // Update user extension document
+              const updateQuery = { programRoleMapping: existingUserProgramRoleMapping };
+              user = await userExtensionsQueries.updateUserExtensionDocument({ _id: existingUser._id }, updateQuery, {
+                new: true,
+                returnNewDocument: true,
+              });
+              userExtensionMap[user.userId.toString()] = user;
 
               userRole['_SYSTEM_ID'] = existingUser._id;
               userRole.status = 'Success';
-            }
-
-            let entityObject = {};
 
-            if (removeUserFromEntity) {
-              entityObject.entityId = userRole.entity;
-              entityObject.role = userRole.role;
+              userRole._kafkaEventPayloads = kafkaEventPayloads;
+              aggregateKafkaEventPayloads.push(...kafkaEventPayloads);
             }
 
-            await this.pushUserToElasticSearch(user._doc, entityObject);
           } catch (error) {
+            console.log(error, 'error in user role creation');
```

Comment:
@borkarsaish65 Please remove the consoles

Reply:
done